File: problem-solving/_lc-0188-hard-best-time-to-buy-and-sell-stock-iv/2-bottom-up.go
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/

Complexity
----------
Time: O(k*n)
Space: O(k)

Reference
---------
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/solutions/54125/very-understandable-solution-by-reusing-problem-iii-idea/comments/1243324
*/

func maxProfit(k int, prices []int) int {
    var buy []int
    profit := make([]int, k + 1)
    for idx := 0; idx <= k; idx++ {
        buy = append(buy, math.MaxInt32)
    }
    for p := 0; p < len(prices); p++ {
        for i := 1; i <= k; i++ {
            buy[i] = min(buy[i], prices[p] - profit[i - 1])
            profit[i] = max(profit[i], prices[p] - buy[i])
        }
    }
    return profit[k]
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: problem-solving/_lc-123-hard-best-time-to-buy-and-sell-stock-iii/2-bottom-up.py
"""
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/

Complexity
----------
Time: O(n)
Space: O(n)

Reference
---------
Approach-3: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/solutions/796990/c-worst-to-best-solution-explained-for-dummies-like-me/
"""

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy1 = buy2 = math.inf
        profit1 = profit2 = -math.inf
        for i, p in enumerate(prices):
            buy1 = min(buy1, p)
            profit1 = max(profit1, p - buy1)
            buy2 = min(buy2, p - profit1)
            profit2 = max(profit2, p - buy2)
        return profit2
